windows/src/engine/inst-cpp/insthelper.cpp

```diff
@@ -0,0 +1,211 @@
+  std::wostringstream result;
+
+  result << std::hex << std::setw(4) << std::setfill(L'0') << LangID << L":" << GuidToWString(c_clsidKMTipTextService)
+         << GuidToWString(guidProfile);
+
+  return result.str();
+}
+
+bool
+InstallLayoutOrTip(const wchar_t* FLayoutInstallString, DWORD Flags) {
+  HMODULE hInputDll;
+  BOOL result = FALSE;
+
+  // Function pointer type for InstallLayoutOrTip
+  typedef BOOL(WINAPI * TInstallLayoutOrTipFunc)(const wchar_t*, DWORD);
+
+  // Load the DLL
+  hInputDll = LoadLibrary(L"input.dll");
+  if (!hInputDll) {
+    return false;
+  }
+
+  // Get the function pointer
+  TInstallLayoutOrTipFunc PInstallLayoutOrTip;
+  PInstallLayoutOrTip = (TInstallLayoutOrTipFunc)GetProcAddress(hInputDll, "InstallLayoutOrTip");
+  if (!PInstallLayoutOrTip) {
+    FreeLibrary(hInputDll);
+    return false;
+  }
+
+  // Call the function
+  result = PInstallLayoutOrTip(FLayoutInstallString, Flags);
+
+  // Free the DLL
+  FreeLibrary(hInputDll);
+
+  return result ? true : false;
+}
+// Unregister the TIP
+void UnregisterTIPAndItsProfiles(const CLSID& AClsid) {
+  HRESULT hr;
+  CComPtr<ITfInputProcessorProfiles> pInputProcessorProfiles;
+  CComPtr<ITfInputProcessorProfileMgr> pInputProcessorProfileMgr;
+  CComPtr<IEnumTfInputProcessorProfiles> ippEnum;
+  ULONG pcFetch;
+  TF_INPUTPROCESSORPROFILE profile;
+  std::wstring FLayoutInstallString;
+
+  // Create an instance of ITF Input Processor Profiles
+  hr = CoCreateInstance(
+      CLASS_TF_InputProcessorProfiles, NULL, CLSCTX_INPROC_SERVER, IID_ITfInputProcessorProfiles,
+      (void**)&pInputProcessorProfiles);
+  if (FAILED(hr))
+    throw _com_error(hr);
+
+  // Query for ITfInputProcessorProfileMgr interface
+  hr = pInputProcessorProfiles->QueryInterface(IID_ITfInputProcessorProfileMgr, (void**)&pInputProcessorProfileMgr);
+  if (FAILED(hr))
+    throw _com_error(hr);
+
+  // Enumerate the profiles
+  hr = pInputProcessorProfileMgr->EnumProfiles(0, &ippEnum);
+  if (FAILED(hr))
+    throw _com_error(hr);
+
+  // Unregister the input profiles installed through Keyman
+  while (ippEnum->Next(1, &profile, &pcFetch) == S_OK) {
+    if (profile.dwProfileType == TF_PROFILETYPE_INPUTPROCESSOR && IsEqualGUID(profile.clsid, AClsid)) {
+      FLayoutInstallString = GetLayoutInstallString(profile.langid, profile.guidProfile);
+      InstallLayoutOrTip(FLayoutInstallString.c_str(), ILOT_UNINSTALL);
+      pInputProcessorProfileMgr->UnregisterProfile(AClsid, profile.langid, profile.guidProfile, 0);
+    }
+  }
+}
+
+extern "C" __declspec(dllexport) unsigned int PreUninstall() {
+  // Initialize COM
+  HRESULT hr = CoInitializeEx(NULL, COINIT_APARTMENTTHREADED);
+  if (SUCCEEDED(hr)) {
+    try {
+      UnregisterTIPAndItsProfiles(c_clsidKMTipTextService);
+    } catch (...) {
+      // Swallow exceptions so we don't break the uninstall
+    }
+
+    // Clean up COM
+    CoUninitialize();
+  }
+  else
+    throw _com_error(hr);
```

Comment:
We mustn't throw an exception across an export boundary. We need to return an error code instead.